Documentation/ceph-csi-troubleshooting.md

```diff
@@ -0,0 +1,388 @@
+
+If you see a driver not found an error in the application pod describe output.
+Restarting the `csi-xxxxplugin-xxx` pod on the node may help.
+
+## Volume Attachment
+
+Each provisioner pod also has a sidecar container called `csi-attacher`.
+
+### csi-attacher
+
+The external-attacher is a sidecar container that attaches volumes to nodes by calling `ControllerPublish` and
+`ControllerUnpublish` functions of CSI drivers. It is necessary because the internal Attach/Detach controller
+running in Kubernetes controller-manager does not have any direct interfaces to CSI drivers. More details can
+be found [here](https://github.com/kubernetes-csi/external-attacher).
+
+If any issue exists in attaching the PVC to the application pod first check the volumettachment object created
+and also log from csi-attacher sidecar container in provisioner pod.
+
+```console
+$ kubectl get volumeattachment
+NAME                                                                   ATTACHER                        PV                                         NODE       ATTACHED   AGE
+csi-75903d8a902744853900d188f12137ea1cafb6c6f922ebc1c116fd58e950fc92   rook-ceph.cephfs.csi.ceph.com   pvc-5c547d2a-fdb8-4cb2-b7fe-e0f30b88d454   minikube   true       4m26s
+
+$ kubectl logs po/csi-rbdplugin-provisioner-d857bfb5f-ddctl -c csi-attacher
+```
+
+## CephFS Stale operations
+
+Check for any stale mount commands on the `csi-cephfsplugin-xxxx` pod on the node where your application pod is scheduled.
+
+You need to exec in the `csi-cephfsplugin-xxxx` pod and grep for stale mount operators.
+
+Identify the `csi-cephfsplugin-xxxx` pod running on the node where your application is scheduled with
+`kubectl get po -o wide` and match the node names.
+
+```console
+$ kubectl exec -it csi-cephfsplugin-tfk2g -c csi-cephfsplugin -- sh
+sh-4.4# ps -ef |grep mount
+root          67      60  0 11:55 pts/0    00:00:00 grep mount
+sh-4.4# ps -ef |grep ceph
+root           1       0  0 Jan20 ?        00:00:26 /usr/local/bin/cephcsi --nodeid=minikube --type=cephfs --endpoint=unix:///csi/csi.sock --v=0 --nodeserver=true --drivername=rook-ceph.cephfs.csi.ceph.com --pidlimit=-1 --metricsport=9091 --forcecephkernelclient=true --metricspath=/metrics --enablegrpcmetrics=true
+root          69      60  0 11:55 pts/0    00:00:00 grep ceph
+```
+
+If any commands are stuck check the **dmesg** logs from the node.
+Restarting the `csi-cephfsplugin` pod may also help sometimes.
+
+If you don’t see any stuck messages, confirm the network connectivity, Ceph health, and slow ops.
+
+## RBD Stale operations
+
+Check for any stale `map/mkfs/mount` commands on the `csi-rbdplugin-xxxx` pod on the node where your application pod is scheduled.
+
+You need to exec in the `csi-rbdplugin-xxxx` pod and grep for stale operators like (`rbd map, rbd unmap, mkfs, mount` and `umount`).
+
+Identify the `csi-rbdplugin-xxxx` pod running on the node where your application is scheduled with
```

Comment:
Use the deployment instead.

Reply:
deployment is not helpful here as daemonset pods are responsible for mounting the PVC.